adserver/reports.py

```diff
@@ -0,0 +1,283 @@
+from .utils import calculate_ctr
+from .utils import calculate_ecpm
+from .utils import get_country_name
+
+
+log = logging.getLogger(__name__)  # noqa
+
+
+class BaseReport:
+
+    """Base report which other reports are extended from."""
+
+    # Model of the aggregated impression class
+    # This is used to validate that the expected queryset class matches the passed one
+    model = None
+
+    # This is a "getter" (potentially dotted) used to get the field that the data is indexed by
+    # For example, to get a report broken down by day, use ``date`` which will get ``adimpression.date``
+    index = None
+
+    # The "getter" to order the results by (``-views`` to order by views descending)
+    order = None
+
+    DEFAULT_MAX_RESULTS = 65535
+
+    # Fields to apply select_related on the queryset
+    select_related_fields = ("advertisement", "advertisement__flight")
+
+    def __init__(self, queryset, index=None, order=None, max_results=None, **kwargs):
+        """
+        Initialize the report using filtered, ordered queryset.
+
+        :param queryset: A filtered queryset to use with the report.
+        :param index: Override the report result index.
+            This is sometimes necessary for example when filtering a geo report to a specific country,
+            you probably want to index by date.
+        :param order: Override the report result order.
+        :param max_results: Override the maximum results to return
+        """
+        self.queryset = queryset
+
+        if index:
+            self.index = index
+        if order:
+            self.order = order
+        if max_results:
+            self.max_results = max_results
+        else:
+            self.max_results = self.DEFAULT_MAX_RESULTS
+
+        # Save any other keyword args for use in subclasses
+        self.kwargs = kwargs
+
+        # Processed fields available after the report has been generated
+        self.total = {}
+        self.results = []
+
+        if self.queryset.model is not self.model:
+            raise RuntimeError(
+                f"Report queryset (type {self.queryset.model}) is not type {self.model}"
+            )
+
+    def get_index_display(self, index):
+        """Used to add display logic the index field."""
+        return index
+
+    def generate(self):
+        raise NotImplementedError("Subclasses implement this method")
+
+
+class AdvertiserReport(BaseReport):
+
+    """Report for showing daily ad performance for an advertiser."""
+
+    model = AdImpression
+    index = "date"
+    order = "-index"
+
+    def generate(self):
+        """Generate/calculate the report from the queryset by the index."""
+        queryset = self.queryset.select_related(*self.select_related_fields)
+
+        # This allows us to use `.` in the report_index to span relations
+        getter = operator.attrgetter(self.index)
+
+        results = {}
+
+        for impression in queryset:
+            index = getter(impression)
```

Comment:
This breaks on the advertiser report when there isn't an ad:

```
'NoneType' object has no attribute 'flight'
```

That queryset needs to filter `advertisement__isnull=False`:

https://github.com/readthedocs/ethical-ad-server/pull/291/files/c7ce6790cdf610c3ba8e3c35dd81c9fe151db3c8#diff-4e0bd0349e45aaa7670924cf9fc0e3dd183aebe0eaf5bce287975128dbb95509L252